src/core/thread/mesh_forwarder.cpp

```diff
@@ -306,6 +306,11 @@ void MeshForwarder::RemoveMessages(Child &aChild, uint8_t aSubType)
         {
             message->ClearChildMask(childIndex);
             mSourceMatchController.DecrementMessageCount(aChild);
+
+            if (aChild.GetIndirectMessage() == message)
+            {
+                aChild.SetIndirectMessage(NULL);
+            }
```

Comment:
How about creating a function RemoveMessageFromChild(Child &aChild, Message &aMessage); which performs all these operations and can be called from RemoveMessage()? and RemoveMessages()?

Reply:
Good idea.